File: LHiOSAccumulatesInObjectiveC/LHiOSAccumulatesInObjectiveC/General/LHHelpers/UIView+LHAutoLayout.m
//
//  UIView+LHAutoLayout.m
//  jiaxiaozhijia-ios
//
//  Created by 李辉 on 2017/5/5.
//
//

#import "UIView+LHAutoLayout.h"

@implementation UIView (LHAutoLayout)

- (void)insertSubviewUsingDefaultLayoutConstraints:(UIView *)view atIndex:(NSInteger)index {
    [self insertSubview:view atIndex:index];
    [self addDefaultConstraintsForSubView:view];
}

- (void)insertSubviewUsingDefaultLayoutConstraints:(UIView *)view aboveSubview:(UIView *)siblingSubview {
    [self insertSubview:view aboveSubview:siblingSubview];
    [self addDefaultConstraintsForSubView:view];
}

- (void)insertSubviewUsingDefaultLayoutConstraints:(UIView *)view belowSubview:(UIView *)siblingSubview {
    [self insertSubview:view belowSubview:siblingSubview];
    [self addDefaultConstraintsForSubView:view];
}

- (void)addSubviewUsingDefaultLayoutConstraints:(UIView *)view {
    [self addSubview:view];
    [self addDefaultConstraintsForSubView:view];
}

- (void)addDefaultConstraintsForSubView:(UIView *)view {
    view.translatesAutoresizingMaskIntoConstraints = NO;
    [self addConstraints:[NSLayoutConstraint constraintsWithVisualFormat:@"H:|-0-[view]-0-|" options:0 metrics:nil views:NSDictionaryOfVariableBindings(view)]];
    [self addConstraints:[NSLayoutConstraint constraintsWithVisualFormat:@"V:|-0-[view]-0-|" options:0 metrics:nil views:NSDictionaryOfVariableBindings(view)]];
}

- (void)addSubview:(UIView *)view withLayoutInfo:(LHLayoutInfo)info {
    if (!view) {
        return;
    }
    [self addSubview:view];
    view.translatesAutoresizingMaskIntoConstraints = NO;
    [self layoutView:view constant:info.top name:NSLayoutAttributeTop];
    [self layoutView:view constant:info.left name:NSLayoutAttributeLeft];
    [self layoutView:view constant:info.bottom name:NSLayoutAttributeBottom];
    [self layoutView:view constant:info.right name:NSLayoutAttributeRight];
    [self layoutView:view constant:info.width name:NSLayoutAttributeWidth];
    [self layoutView:view constant:info.height name:NSLayoutAttributeHeight];
}

- (void)layoutView:(UIView *)view constant:(CGFloat)constant name:(NSLayoutAttribute)attribute {
    if (constant != LHLayoutNone) {
        NSString *vfl = nil;
        NSDictionary *metrics = @{@"constant": @(constant)};
        switch (attribute) {
            case NSLayoutAttributeTop:
                vfl = @"V:|-constant-[view]";
                break;
            case NSLayoutAttributeLeft:
                vfl = @"H:|-constant-[view]";
                break;
            case NSLayoutAttributeBottom:
                vfl = @"V:[view]-constant-|";
                break;
            case NSLayoutAttributeRight:
                vfl = @"H:[view]-constant-|";
                break;
            case NSLayoutAttributeWidth:
                vfl = @"H:[view(constant)]";
                break;
            case NSLayoutAttributeHeight:
                vfl = @"V:[view(constant)]";
                break;
                
            default:
                break;
        }
        if (vfl) {
            [self addConstraints:[NSLayoutConstraint constraintsWithVisualFormat:vfl options:0 metrics:metrics views:NSDictionaryOfVariableBindings(view)]];
        }
    }
}

- (NSLayoutConstraint *)lh_widthConstraint {
    for (NSLayoutConstraint *constraint in self.constraints) {
        if (constraint.firstItem == self && constraint.firstAttribute == NSLayoutAttributeWidth) {
            return constraint;
        }
    }
    return nil;
}

- (NSLayoutConstraint *)lh_heightConstraint {
    for (NSLayoutConstraint *constraint in self.constraints) {
        if (constraint.firstItem == self && constraint.firstAttribute == NSLayoutAttributeHeight) {
            return constraint;
        }
    }
    return nil;
}

@end

File: LHiOSAccumulatesInSwift/LHiOSAccumulatesInSwift/LSGenreal/LHUtils/LHExtensions/UIView+LHAutoLayout.swift
//
//  UIView+LHAutoLayout.swift
//  LHiOSAccumulatesInSwift
//
//  Created by 李辉 on 2017/7/11.
//  Copyright © 2017年 Lihux. All rights reserved.
//

import UIKit

struct LHLayoutInfo {
    init(top: CGFloat, left: CGFloat, bottom: CGFloat, right: CGFloat, width: CGFloat, height: CGFloat) {
        self.top = top
        self.left = left
        self.bottom = bottom
        self.right = right
        self.width = width
        self.height = height
    }
    var top: CGFloat
    var left: CGFloat
    var bottom: CGFloat
    var right: CGFloat
    var width: CGFloat
    var height: CGFloat
    static let layoutNone = CGFloat.greatestFiniteMagnitude
}

extension UIView {
    func addSubViewUsingDefaultLayoutConstraints(_ view: UIView) -> Void {
        self.addSubview(view)
        self.addDefaultConstraintsForSubView(view)
    }
    
    func addSubView(_ view: UIView, layoutInfo:LHLayoutInfo) -> Void {
        self.addSubview(view)
        view.translatesAutoresizingMaskIntoConstraints = false
        self.layoutView(view, constant: layoutInfo.top, attributeName: .top)
        self.layoutView(view, constant: layoutInfo.left, attributeName: .left)
        self.layoutView(view, constant: layoutInfo.bottom, attributeName: .bottom)
        self.layoutView(view, constant: layoutInfo.right, attributeName: .right)
        self.layoutView(view, constant: layoutInfo.width, attributeName: .width)
        self.layoutView(view, constant: layoutInfo.height, attributeName: .height)
    }
    
    func insertSubviewUsingDefaultLayoutConstraints(_ view: UIView, at:Int) -> Void {
        self.insertSubview(view, at: at)
        self.addSubViewUsingDefaultLayoutConstraints(view)
    }
    
    func insertSubviewUsingDefaultLayoutConstraints(_ view: UIView, aboveSubview: UIView) -> Void {
        self.insertSubview(view, aboveSubview: aboveSubview)
        self.addSubViewUsingDefaultLayoutConstraints(view)
    }
    
    func insertSubviewUsingDefaultLayoutConstraints(_ view: UIView, belowSubview: UIView) -> Void {
        self.insertSubview(view, belowSubview: belowSubview)
        self.addSubViewUsingDefaultLayoutConstraints(view)
    }
    
    func lh_widthConstraint() -> NSLayoutConstraint? {
        var result: NSLayoutConstraint?
        self.constraints.forEach { (constraint: NSLayoutConstraint) in
            if constraint.firstItem as! NSObject == self && constraint.firstAttribute == .width {
                result = constraint
            }
        }
        return result
    }
    func lh_heightConstraint() -> NSLayoutConstraint? {
        var result: NSLayoutConstraint?
        self.constraints.forEach { (constraint: NSLayoutConstraint) in
            if constraint.firstItem as! NSObject == self && constraint.firstAttribute == .width {
                result = constraint
            }
        }
        return result
    }
    
    private func layoutView(_ view:UIView, constant: CGFloat, attributeName: NSLayoutAttribute) -> Void {
        if constant != CGFloat.greatestFiniteMagnitude {
            var vfl:String?
            let metrics = ["constant": constant]
            switch attributeName {
            case .top:
                vfl = "V:|-constant-[view]"
            case .left:
                vfl = "H:|-constant-[view]"
            case .bottom:
                vfl = "V:[view]-constant-|"
            case .right:
                vfl = "H:[view]-constant-|"
            case .width:
                vfl = "H:[view(constant)]"
            case .height:
                vfl = "V:[view(constant)]"
            default:
                break
            }
            if let temp = vfl {
                self.addConstraints(NSLayoutConstraint.constraints(withVisualFormat: temp, options: NSLayoutFormatOptions(rawValue: 0), metrics: metrics, views: ["view": view]))
            }
        }
    }
    
    private func addDefaultConstraintsForSubView(_ view: UIView) -> Void {
        view.translatesAutoresizingMaskIntoConstraints = false
        let hConstraints = NSLayoutConstraint.constraints(withVisualFormat: "H:|-0-[view]-0-|", options: .alignAllBottom, metrics: nil, views: ["view": view])
        let vConstraints = NSLayoutConstraint.constraints(withVisualFormat: "V:|-0-[view]-0-|", options: .alignAllBottom, metrics: nil, views: ["view": view])
        self.addConstraints(hConstraints)
        self.addConstraints(vConstraints)
    }
}
